language/vm/vm-runtime/src/system_txn/block_metadata_processor.rs

```diff
@@ -36,27 +39,57 @@ pub(crate) fn process_block_metadata(
     let mut txn_data = TransactionMetadata::default();
 
     txn_data.max_gas_amount = GasUnits::new(std::u64::MAX);
+    let mut interpreter_context =
+        TransactionExecutionContext::new(txn_data.max_gas_amount(), data_cache);
     // TODO: We might need a non zero cost table here so that we can at least bound the execution
     //       time by a reasonable amount.
     let gas_schedule = CostTable::zero();
 
-    // The transaction executor failed to start
-    let mut txn_executor = TransactionExecutor::new(&gas_schedule, data_cache, txn_data);
     let result = if let Ok((id, timestamp, previous_vote, proposer)) = block_metadata.into_inner() {
         let args = vec![
             Value::u64(timestamp),
             Value::byte_array(id),
             Value::byte_array(previous_vote),
             Value::address(proposer),
         ];
-        txn_executor.execute_function(runtime, &LIBRA_SYSTEM_MODULE, &BLOCK_PROLOGUE, args)
+        runtime.execute_function(
+            &mut interpreter_context,
+            &txn_data,
+            &gas_schedule,
+            &LIBRA_SYSTEM_MODULE,
+            &BLOCK_PROLOGUE,
+            args,
+        )
     } else {
         Err(VMStatus::new(StatusCode::MALFORMED))
     };
     result
-        .and_then(|_| txn_executor.make_write_set(Ok(())))
+        .and_then(|_| make_write_set(&mut interpreter_context, &txn_data))
         .and_then(|output| {
             data_cache.push_write_set(output.write_set());
             Ok(output)
         })
 }
+
+fn make_write_set(
```

Comment:
@runtian-zhou I have moved this function from `TransactionExecutor` here.
I am not sure this is the best place but I'll leave it to you to find a better home given you are actively working in this area